lib/std/mem/Allocator.zig

```diff
@@ -112,7 +112,7 @@ pub fn destroy(self: Allocator, ptr: anytype) void {
     const info = @typeInfo(@TypeOf(ptr)).Pointer;
     const T = info.child;
     if (@sizeOf(T) == 0) return;
-    const non_const_ptr = @intToPtr([*]u8, @ptrToInt(ptr));
+    const non_const_ptr = @qualCast([*]u8, @as([*]const u8, @ptrCast([*]const u8, ptr)));
```

Comment:
This example makes the `@as()` look even more ridiculous, but it is in fact necessary to compile:
```
lib/std/mem/Allocator.zig:115:27: error: '@qualCast' can only modify 'const' and 'volatile' qualifiers
    const non_const_ptr = @qualCast([*]u8, @ptrCast([*]const u8, ptr));
                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/std/mem/Allocator.zig:115:27: note: expected type '[*]const u8'
lib/std/mem/Allocator.zig:115:27: note: got type '[*]align(8) const u8'
```
cc @Vexu

Reply:
Both of these should be fixed by making the destination type of `@qualCast` `[*]align(8) u8`. The reason I made `@qualCast` so strict is to keep it simple.

Reply:
Hmm, was there any discussion on what the semantics of `@qualCast` should be? I couldn't find any.

It seems to me that having two separate builtins `@constCast()` and `@volatileCast()` which each take a single argument and return a value of the exact same type but with the respective qualifier removed would be just as simple and much more user friendly in situations like this.

Reply:
> Hmm, was there any discussion on what the semantics of `@qualCast` should be? I couldn't find any.

Just a general acknowledgement of needing one.

> It seems to me that having two separate builtins `@constCast()` and `@volatileCast()` which each take a single argument and return a value of the exact same type but with the respective qualifier removed would be just as simple and much more user friendly in situations like this.

Sure but I think it could also be solved by making the error note show a suggested type instead of the expected and actual type.

Reply:
Fwiw, I think the `constCast` and `volatileCast` described are implementable using status-quo `@qualCast` (and sound like a good addition to `std.meta` to me).

Reply:
I don't have the issue number handy but remember that it is an accepted proposal (that I am quite looking forward to in fact) to remove the destination type from all of the casting builtins. This will include `@qualCast`.

Reply:
Being able to accidentally cast away `const` while trying to cast (away) `volatile` sounds like a **massive** footgun to me, especially with eventual implicit destination types. This is made worse by the fact that `volatile` almost entirely used in embedded, which notoriously has limited debugging facilities. I would be frightened to ever use `@qualCast` without an immediate `@as` to enforce that I'm not casting away `const`, but why make the easier option the dangerous one? More explicit is more better, especially when the cost is almost free. +1 for `@constCast` and `@volatileCast`

Reply:
I'll make the call right now because it's an easy one. 

Replace `@qualCast` with `@constCast` and `@volatileCast` and remove the destination type parameter.